### ISSUE TITLE
Domain Adaptation Losses 

### ISSUE THREAD

**Issue by quizzy5889** [GiovanniPasq/DA-Faster-RCNN#14]
Ciao Giovanni, 
first of all thanks for your work. It's very interesting to have this project ported to Pytorch 2.x!!

Mine is mainly a curiosity. I was having a look at the losses trend in the notebook and to me it seems that da instance loss and da image loss are decreasing very slowly if compared to the others. 

Is this somehow expected?
How can we guarantee that the network is actually learning also the adaptation part?

Grazie

**Comment by GiovanniPasq** [comment]
Hello,

The loss trend is influenced by the dataset, discriminator architecture, and the alpha hyperparameter, which balances the gradient return. In the notebook, I set this value to 0.5; you can experiment by increasing it to observe a different loss trend (the maximum value is 1). Alternatively, you can modify the discriminator architectures.

Best regards,
Giovanni.

**Comment by quizzy5889** [comment]
Thanks for your answer, I really appreciate.

 So for example I'm trying to train with my own dataset and in this moment I have a loss_instance_d which is decreasing, while loss_image_d is somehow still as you can see here:

<img width="350" alt="Screenshot 2024-01-19 at 14 29 13" src="https://github.com/GiovanniPasq/DA-Faster-RCNN/assets/3774189/21b5dd40-d8ff-451f-a9e7-526fb86c4b07">

What would you suggest to do? Because modifying alpha will affect both da_losses, right? Non just loss_image_d...

**Comment by GiovanniPasq** [comment]
Setting the 0.5 to 1 this will effect both discriminators.
```
loss_dict = model(data_source, False, 0.5)
loss_dict_target = model(data_target, True, 0.5)
```
to
```
loss_dict = model(data_source, False, 1)
loss_dict_target = model(data_target, True, 1)
```

if you want to change only one of them you can just edit the discriminator you want by changing this line with the value that you want to test  ``` x = GradReverse.apply(x, alpha) ``` -> ``` x = GradReverse.apply(x, 1) ```

Otherwise, as i suggested, you can edit the discriminator with your custom network. It is important to notice that does not exist the "perfect" discriminator for every dataset and that the loss behaviour change for each of them. If the loss only decrease does not mean that the network is not learning to adapt. It is altro true that the behaviour of these losses should be similar to the behaviour of the GAN loss.
The same author that introduced the "Gradient reversal layer" have created 3 different discriminator for each of the 3 dataset that they use for their tests.
Check the link, page 10 https://arxiv.org/pdf/1409.7495.pdf

To check if the network is improving their performance you should do the following test:
1) train with the labeled source the standard Faster RCNN and test directly on the target test set.
2) train with the labeled source and unlabeled target DA Faster RCNN and test on the target test set.
3) compare the mAP of Faster RCNN and DA-Faster RCNN. If the mAP of DA-Faster RCNN is higher the network is learning to adapt.

Best regards,
Giovanni
